File: prayer_app_flutter/lib/src/screens/saved_azkar_screen.dart
import 'dart:convert';
import 'package:flutter/material.dart';
import 'package:shared_preferences/shared_preferences.dart';
import '../theme/app_theme.dart';
import '../providers/theme_provider.dart';
import '../models/azkar_data.dart';
import 'azkar_detail_screen.dart';

class SavedAzkarScreen extends StatefulWidget {
  const SavedAzkarScreen({super.key});

  @override
  State<SavedAzkarScreen> createState() => _SavedAzkarScreenState();
}

class _SavedAzkarScreenState extends State<SavedAzkarScreen> {
  List<Map<String, dynamic>> _favorites = [];

  @override
  void initState() {
    super.initState();
    _loadFavorites();
  }

  Future<void> _loadFavorites() async {
    final prefs = await SharedPreferences.getInstance();
    final raw = prefs.getString('azkar_favorites_v1');
    if (raw != null && mounted) {
      setState(() {
        _favorites = (jsonDecode(raw) as List).cast<Map<String, dynamic>>();
      });
    }
  }

  Future<void> _removeFavorite(int listIndex) async {
    final prefs = await SharedPreferences.getInstance();
    setState(() => _favorites.removeAt(listIndex));
    await prefs.setString('azkar_favorites_v1', jsonEncode(_favorites));
  }

  AzkarCategory? _findCategory(String categoryId) {
    try {
      return azkarCategories.firstWhere((c) => c.id == categoryId);
    } catch (_) {
      return null;
    }
  }

  @override
  Widget build(BuildContext context) {
    final tc = ThemeScope.of(context).current;

    return Scaffold(
      body: Container(
        decoration: BoxDecoration(gradient: appBackgroundGradient(tc)),
        child: SafeArea(
          bottom: false,
          child: Column(
            children: [
              // Top bar
              Padding(
                padding: const EdgeInsets.symmetric(horizontal: 8, vertical: 8),
                child: Row(
                  children: [
                    IconButton(
                      icon: Icon(Icons.arrow_back_ios, color: tc.textPrimary, size: 20),
                      onPressed: () => Navigator.of(context).pop(),
                    ),
                    Expanded(
                      child: Text(
                        'Saved Azkar',
                        style: AppTypography.titleMedium(tc),
                        textAlign: TextAlign.center,
                      ),
                    ),
                    const SizedBox(width: 48), // balance
                  ],
                ),
              ),
              // Content
              Expanded(
                child: _favorites.isEmpty
                    ? Center(
                        child: Column(
                          mainAxisSize: MainAxisSize.min,
                          children: [
                            Icon(Icons.bookmark_outline, size: 48, color: tc.textMuted),
                            const SizedBox(height: 12),
                            Text(
                              'No saved azkar yet.',
                              style: TextStyle(
                                fontFamily: 'Inter',
                                fontSize: 15,
                                color: tc.textMuted,
                              ),
                            ),
                          ],
                        ),
                      )
                    : ListView.separated(
                        padding: EdgeInsets.only(
                          left: AzkarLayout.screenPadding,
                          right: AzkarLayout.screenPadding,
                          bottom: MediaQuery.of(context).padding.bottom + AzkarLayout.footerBottomInset,
                        ),
                        itemCount: _favorites.length,
                        separatorBuilder: (_, __) => SizedBox(height: AzkarLayout.listCardSpacing),
                        itemBuilder: (context, i) {
                          final fav = _favorites[i];
                          final catId = fav['categoryId'] as String;
                          final idx = fav['index'] as int;
                          final cat = _findCategory(catId);
                          if (cat == null || idx >= cat.items.length) {
                            return const SizedBox.shrink();
                          }
                          final item = cat.items[idx];
                          final preview = item.arabic.replaceAll('\n', ' ');
                          final previewText = preview.length > 60
                              ? '${preview.substring(0, 60)}…'
                              : preview;

                          return GestureDetector(
                            onTap: () {
                              Navigator.of(context).push(
                                MaterialPageRoute(
                                  builder: (_) => AzkarDetailScreen(
                                    category: cat,
                                    initialIndex: idx,
                                  ),
                                ),
                              ).then((_) => _loadFavorites());
                            },
                            child: Container(
                              padding: EdgeInsets.all(AzkarLayout.listCardPadding),
                              decoration: BoxDecoration(
                                color: tc.card,
                                borderRadius: BorderRadius.circular(AzkarLayout.gridCardRadius),
                                border: Border.all(color: tc.cardBorder),
                              ),
                              child: Row(
                                children: [
                                  Expanded(
                                    child: Column(
                                      crossAxisAlignment: CrossAxisAlignment.start,
                                      children: [
                                        Text(
                                          cat.title,
                                          style: TextStyle(
                                            fontFamily: 'Inter',
                                            fontSize: 11,
                                            color: tc.textMuted,
                                          ),
                                        ),
                                        const SizedBox(height: 6),
                                        Text(
                                          previewText,
                                          style: TextStyle(
                                            fontSize: 15,
                                            color: tc.textPrimary,
                                            height: 1.6,
                                          ),
                                          textDirection: TextDirection.rtl,
                                          maxLines: 2,
                                          overflow: TextOverflow.ellipsis,
                                        ),
                                        if (item.translation.isNotEmpty) ...[
                                          const SizedBox(height: 4),
                                          Text(
                                            item.translation,
                                            style: TextStyle(
                                              fontFamily: 'Inter',
                                              fontSize: 12,
                                              color: tc.textMuted,
                                            ),
                                          ),
                                        ],
                                      ],
                                    ),
                                  ),
                                  const SizedBox(width: 12),
                                  GestureDetector(
                                    onTap: () => _removeFavorite(i),
                                    child: Icon(Icons.bookmark, size: 22, color: tc.accent),
                                  ),
                                ],
                              ),
                            ),
                          );
                        },
                      ),
              ),
            ],
          ),
        ),
      ),
    );
  }
}

File: prayer_app_expo/src/screens/SavedAzkarScreen.js
import React, { useState, useEffect, useCallback } from 'react';
import { View, Text, TouchableOpacity, FlatList, StyleSheet } from 'react-native';
import AsyncStorage from '@react-native-async-storage/async-storage';
import Icon from 'react-native-vector-icons/MaterialCommunityIcons';
import { useTheme } from '../providers/ThemeProvider';
import { AzkarLayout } from '../theme/theme';
import { getTypography, interFont } from '../theme/theme';
import { azkarCategories } from '../data/azkarData';
import { useSafeAreaInsets } from 'react-native-safe-area-context';

export default function SavedAzkarScreen({ onBack, onOpenItem }) {
    const { theme: tc } = useTheme();
    const typo = getTypography(tc);
    const insets = useSafeAreaInsets();
    const [favorites, setFavorites] = useState([]);

    const loadFavorites = useCallback(async () => {
        try {
            const raw = await AsyncStorage.getItem('azkar_favorites_v1');
            if (raw) setFavorites(JSON.parse(raw));
            else setFavorites([]);
        } catch (_) { setFavorites([]); }
    }, []);

    useEffect(() => { loadFavorites(); }, [loadFavorites]);

    const removeFavorite = useCallback(async (index) => {
        const updated = [...favorites];
        updated.splice(index, 1);
        setFavorites(updated);
        try {
            await AsyncStorage.setItem('azkar_favorites_v1', JSON.stringify(updated));
        } catch (_) { /* skip */ }
    }, [favorites]);

    const findCategory = (categoryId) => {
        return azkarCategories.find(c => c.id === categoryId) || null;
    };

    const renderItem = ({ item: fav, index: listIdx }) => {
        const cat = findCategory(fav.categoryId);
        if (!cat || fav.index >= cat.items.length) return null;
        const azkarItem = cat.items[fav.index];
        const preview = azkarItem.arabic.replace(/\n/g, ' ');
        const previewText = preview.length > 60 ? preview.substring(0, 60) + '…' : preview;

        return (
            <TouchableOpacity
                activeOpacity={0.7}
                onPress={() => onOpenItem(cat, fav.index)}
                style={[styles.itemCard, {
                    backgroundColor: tc.card,
                    borderColor: tc.cardBorder,
                }]}
            >
                <View style={{ flex: 1 }}>
                    <Text style={[styles.catLabel, { color: tc.textMuted }]}>{cat.title}</Text>
                    <View style={{ height: 6 }} />
                    <Text
                        style={[styles.previewText, { color: tc.textPrimary }]}
                        numberOfLines={2}
                    >{previewText}</Text>
                    {azkarItem.translation ? (
                        <>
                            <View style={{ height: 4 }} />
                            <Text style={[styles.translation, { color: tc.textMuted }]}>
                                {azkarItem.translation}
                            </Text>
                        </>
                    ) : null}
                </View>
                <View style={{ width: 12 }} />
                <TouchableOpacity onPress={() => removeFavorite(listIdx)} hitSlop={8}>
                    <Icon name="bookmark" size={22} color={tc.accent} />
                </TouchableOpacity>
            </TouchableOpacity>
        );
    };

    return (
        <View style={styles.container}>
            {/* Top bar */}
            <View style={styles.topBar}>
                <TouchableOpacity onPress={onBack} hitSlop={12}>
                    <Icon name="chevron-left" size={28} color={tc.textPrimary} />
                </TouchableOpacity>
                <Text style={[typo.titleMedium, { flex: 1, textAlign: 'center' }]}>
                    Saved Azkar
                </Text>
                <View style={{ width: 48 }} />
            </View>

            {favorites.length === 0 ? (
                <View style={styles.empty}>
                    <Icon name="bookmark-outline" size={48} color={tc.textMuted} />
                    <View style={{ height: 12 }} />
                    <Text style={[styles.emptyText, { color: tc.textMuted }]}>
                        No saved azkar yet.
                    </Text>
                </View>
            ) : (
                <FlatList
                    data={favorites}
                    keyExtractor={(item, i) => `${item.categoryId}:${item.index}:${i}`}
                    renderItem={renderItem}
                    contentContainerStyle={{
                        paddingHorizontal: AzkarLayout.screenPadding,
                        paddingBottom: insets.bottom + AzkarLayout.footerBottomInset,
                    }}
                    ItemSeparatorComponent={() => <View style={{ height: AzkarLayout.listCardSpacing }} />}
                />
            )}
        </View>
    );
}

const styles = StyleSheet.create({
    container: { flex: 1 },
    topBar: {
        flexDirection: 'row',
        alignItems: 'center',
        paddingHorizontal: 8,
        paddingVertical: 8,
        gap: 8,
    },
    empty: {
        flex: 1,
        justifyContent: 'center',
        alignItems: 'center',
    },
    emptyText: {
        fontFamily: interFont('400'),
        fontSize: 15,
    },
    itemCard: {
        padding: AzkarLayout.listCardPadding,
        borderRadius: AzkarLayout.gridCardRadius,
        borderWidth: 1,
        flexDirection: 'row',
        alignItems: 'center',
    },
    catLabel: {
        fontFamily: interFont('400'),
        fontSize: 11,
    },
    previewText: {
        fontSize: 15,
        lineHeight: 15 * 1.6,
        writingDirection: 'rtl',
        textAlign: 'right',
    },
    translation: {
        fontFamily: interFont('400'),
        fontSize: 12,
    },
});
